tap_rest_api_msdk/streams.py

```diff
@@ -1,13 +1,184 @@
 """Stream type classes for tap-rest-api-msdk."""
 
 from datetime import datetime
-from typing import Any, Dict, Iterable, Optional
+from typing import Any, Dict, Iterable, Optional, cast
 
 import requests
+import email.utils
+from singer_sdk.pagination import SinglePagePaginator, BaseOffsetPaginator, BaseHATEOASPaginator, JSONPathPaginator, HeaderLinkPaginator, SimpleHeaderPaginator, BasePageNumberPaginator
 from singer_sdk.helpers.jsonpath import extract_jsonpath
 from tap_rest_api_msdk.client import RestApiStream
 from tap_rest_api_msdk.utils import flatten_json
-from urllib.parse import urlparse, parse_qsl
+from urllib.parse import urlparse, parse_qsl, parse_qs
+
+
+class RestAPIBasePageNumberPaginator(BasePageNumberPaginator):
+    def __init__(
+        self,
+        *args,
+        jsonpath: str = None,
+        **kwargs
+    ):
+        super().__init__(*args, **kwargs)
+        self._jsonpath = jsonpath
+
+    def has_more(self, response: requests.Response):
+        """Return True if there are more pages to fetch.
+
+        Args:
+            response: The most recent response object.
+            jsonpath: An optional jsonpath to where the tokens are located in
+                      the response, defaults to `hasMore` in the response.
+
+        Returns:
+            Whether there are more pages to fetch.
+        """
+        
+        if self._jsonpath:
+            return next(extract_jsonpath(self._jsonpath, response.json()), None)
+        else:
+            return response.json().get("hasMore", None)
+
+class RestAPIOffsetPaginator(BaseOffsetPaginator):
+    def __init__(
+        self,
+        *args,
+        jsonpath: str = None,
+        **kwargs
+    ):
+        super().__init__(*args, **kwargs)
+        self._jsonpath = jsonpath
+
+    def has_more(self, response: requests.Response):
+        """Return True if there are more pages to fetch.
+
+        Args:
+            response: The most recent response object.
+            jsonpath: An optional jsonpath to where the tokens are located in
+                      the response, defaults to pagination in the response.
+
+        Returns:
+            Whether there are more pages to fetch.
+        """
+        
+        if self._jsonpath:
+            pagination = next(extract_jsonpath(self._jsonpath, response.json()), None)
+        else:
+            pagination = response.json().get("pagination", None)
+
+        if pagination and all(x in pagination for x in ["offset", "limit"]):
+            record_limit = pagination.get("total",pagination.get("count",0))
+            records_read = pagination["offset"] + pagination["limit"]
+            if records_read <= record_limit:
```

Comment:
I was wondering if it would be possible for this always to evaluate to true, however I think offset would shift forward by the page_size and make `records_read` exceed the limit